staging/src/k8s.io/apiserver/pkg/storage/etcd3/etcdfeature/feature_support_checker.go

```diff
@@ -0,0 +1,149 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package etcdfeature
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"k8s.io/apimachinery/pkg/util/version"
+	"k8s.io/apiserver/pkg/storage"
+)
+
+var (
+	// Define these static versions to use for checking version of etcd, issue on kubernetes #123192
+	v3_4_25 = version.MustParseSemantic("3.4.25")
+	v3_5_0  = version.MustParseSemantic("3.5.0")
+	v3_5_8  = version.MustParseSemantic("3.5.8")
+
+	// DefaultFeatureSupportChecker is a shared global FeatureSupportChecker.
+	DefaultFeatureSupportChecker FeatureSupportChecker = newDefaultFeatureSupportChecker()
+)
+
+// FeatureSupportChecker to define Supports functions.
+type FeatureSupportChecker interface {
+	Supports(feature storage.Feature) (bool, error)
+	// This method works with etcd client to recalcuate feature support and caches it.
+	EndpointSupports(ctx context.Context, c client, endpoint string, feature storage.Feature) (bool, error)
+	// Only use for testing and use defer afterwork.
+	SetFeatureSupportCheckerDuringTest(feature storage.Feature, value bool) func()
+}
+
+// featureSupportInfo holds information about feature support status, used for cache.
+type featureSupportInfo struct {
+	Supported bool
+}
+
+type defaultFeatureSupportChecker struct {
+	progresNotifySupport sync.Map
+}
+
+func newDefaultFeatureSupportChecker() *defaultFeatureSupportChecker {
+	return &defaultFeatureSupportChecker{}
+}
+
+// Supports can check the featue from anywhere without storage if it was checked before with Supports
+// Only return true or false if it was defined or not.
+func (f *defaultFeatureSupportChecker) Supports(feature storage.Feature) (bool, error) {
+	switch feature {
+	case storage.RequestWatchProgress:
+		return f.supportsRequestWatchProgress()
+	default:
+		return false, fmt.Errorf("feature %q is not implemented in DefaultFeatureSupportChecker", feature)
+	}
+}
+
+func (f *defaultFeatureSupportChecker) EndpointSupports(ctx context.Context, c client, endpoint string, feature storage.Feature) (bool, error) {
+	// Check if endpoint is cached before or not.
+	if value, ok := f.progresNotifySupport.Load(endpoint); ok {
+		return value.(featureSupportInfo).Supported, nil
+	}
+
+	supported, err := endpointSupportsRequestWatchProgress(ctx, c, endpoint)
+	if err != nil {
+		return false, err
+	}
+	f.progresNotifySupport.Store(feature, featureSupportInfo{
+		Supported: supported,
+	})
+	return supported, nil
+}
+
+func (f *defaultFeatureSupportChecker) supportsRequestWatchProgress() (bool, error) {
+	// use to check after Range.
+	isSupported := true
+
+	f.progresNotifySupport.Range(func(_, value interface{}) bool {
+		info := value.(featureSupportInfo)
+		if !info.Supported {
+			isSupported = false
+			return false
+		}
+		return true
```

Comment:
```suggestion
		isSupported &= info.Supported
		return isSupported
```

Reply:
I think this can't work, maybe `isSupported = info.Supported && isSupported`

Reply:
I tested it, it wont let me compile.

Reply:
Can you provide an error that you get?

Reply:
`invalid operation: operator & not defined on isSupported (variable of type bool)`